Very obscure error when not all subclasses of a sealed class are serializable

Run this code:

```kotlin
import kotlinx.serialization.*
import kotlinx.serialization.json.*

@Serializable
sealed class Repository {
    abstract val name: String
}
            
// THIS IS NOT @Serializable
class SimpleRepository(override val name: String) : Repository()

@Serializable
class OwnedRepository(override val name: String, val owner: String) : Repository()

fun main() {
    val data: Repository = OwnedRepository("kotlinx.coroutines", "kotlin")
    println(Json.encodeToString(data))
}  
```

It results with a very obscure error:

```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at kotlinx.serialization.internal.PlatformKt.constructSerializerForGivenTypeArgs(Platform.kt:40)
	at kotlinx.serialization.internal.PlatformKt.compiledSerializerImpl(Platform.kt:23)
	at kotlinx.serialization.SerializersKt__SerializersKt.serializerOrNull(Serializers.kt:109)
	at kotlinx.serialization.SerializersKt.serializerOrNull(Unknown Source)
	at example.examplePoly04.Example_poly_04Kt.main(example-poly-04.kt:23)
	at example.examplePoly04.Example_poly_04Kt.main(example-poly-04.kt)
Caused by: java.lang.IllegalArgumentException: Arrays of classes and serializers must have the same length, got arrays: [class example.examplePoly04.SimpleRepository, class example.examplePoly04.OwnedRepository], [example.examplePoly04.OwnedRepository$$serializer@281e3708]
Please ensure that @Serializable annotation is present on each sealed subclass
	at kotlinx.serialization.SealedClassSerializer.<init>(SealedSerializer.kt:96)
	at example.examplePoly04.Repository$Companion.serializer(example-poly-04.kt)
	... 10 more
```

Expected: It should either work or, at least, produce a human-friendly error message.

--- a/runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt
+++ b/runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt
@@ -93,11 +93,10 @@ internal class SealedClassSerializer<T : Any>(
     private val serialName2Serializer: Map<String, KSerializer<out T>>
 
     init {
-        require(subclasses.size == subclassSerializers.size) {
-            "Arrays of classes and serializers must have the same length," +
-                    " got arrays: ${subclasses.contentToString()}, ${subclassSerializers.contentToString()}\n" +
-                    "Please ensure that @Serializable annotation is present on each sealed subclass"
+        if (subclasses.size != subclassSerializers.size) {
+            throw IllegalArgumentException("All subclasses of sealed class ${baseClass.simpleName} should be marked @Serializable")
         }
+
         class2Serializer = subclasses.zip(subclassSerializers).toMap()
         serialName2Serializer = class2Serializer.entries.groupingBy { it.value.descriptor.serialName }
             .aggregate<Map.Entry<KClass<out T>, KSerializer<out T>>, String, Map.Entry<KClass<*>, KSerializer<out T>>>

--- a/runtime/jvmMain/src/kotlinx/serialization/internal/Platform.kt
+++ b/runtime/jvmMain/src/kotlinx/serialization/internal/Platform.kt
@@ -30,16 +30,8 @@ internal actual fun <T : Any, E : T?> ArrayList<E>.toNativeArrayImpl(eClass: KCl
 internal actual fun <T : Any> KClass<T>.constructSerializerForGivenTypeArgs(vararg args: KSerializer<Any?>): KSerializer<T>? {
     val jClass = this.java
     // Search for serializer defined on companion object.
-    val companion =
-        jClass.declaredFields.singleOrNull { it.name == "Companion" }?.apply { isAccessible = true }?.get(null)
-    if (companion != null) {
-        val serializer = companion.javaClass.methods
-            .find { method ->
-                method.name == "serializer" && method.parameterTypes.size == args.size && method.parameterTypes.all { it == KSerializer::class.java }
-            }
-            ?.invoke(companion, *args) as? KSerializer<T>
-        if (serializer != null) return serializer
-    }
+   val serializer = invokeSerializerOnCompanion<T>(jClass, *args)
+    if (serializer != null) return serializer
     // Check whether it's serializable object
     findObjectSerializer(jClass)?.let { return it }
     // Search for default serializer if no serializer is defined in companion object.
@@ -53,6 +45,22 @@ internal actual fun <T : Any> KClass<T>.constructSerializerForGivenTypeArgs(vara
     if (default != null) return default
     // Fallback: enum without @Serializable annotation as last resort
     return createEnumSerializer()
+}
+
+private fun <T : Any> invokeSerializerOnCompanion(jClass: Class<*>, vararg args: KSerializer<Any?>): KSerializer<T>? {
+    val companion =
+        jClass.declaredFields.singleOrNull { it.name == "Companion" }?.apply { isAccessible = true }?.get(null)
+            ?: return null
+    try {
+        return companion.javaClass.methods
+            .find { method ->
+                method.name == "serializer" && method.parameterTypes.size == args.size && method.parameterTypes.all { it == KSerializer::class.java }
+            }
+            ?.invoke(companion, *args) as? KSerializer<T>
+    } catch (e: InvocationTargetException) {
+        val cause = e.cause ?: throw e
+        throw InvocationTargetException(cause, cause.message ?: e.message)
+    }
 }
 
 @Suppress("UNCHECKED_CAST")

